[WIP] Iterative optimizer

martint opened the pull request:
This is a preview of an iterative version of the current optimizer framework.

It decouples the traversal of the plan tree (`IterativeOptimizer`) from the transformation logic (`Rule`). It's implemented as yet-another `PlanOptimizer`, so it can be inserted in the middle of the existing optimizers sequence.

Some benefits:
* It makes it possible to write targeted transformation rules that will work even if they are not implemented for every node that might appear in the ancestry of a given node.
* Rules can be tested individually.
* The `Memo` structure breaks up the tree and adds a level of indirection that allows for more efficient mutations.
* It implements column pruning and unaliasing in a more principled way.

It's very much work in progress, but I wanted others to see what's coming up (cc @mattsfuller).

- [ ] Rule tests
- [ ] Migrate predicate pushdown rules
- [ ] Migrate prune outputs rules
- [x] Config & session property
- [x] GC unreferenced nodes in Memo

martint commented:
cc @kokosing

dain commented:
@martint do you want to get this checked in with a flag to enable/disable?

martint commented:
Yes, definitely.

kokosing commented:
I didn't read code end-to-end, so please forgive me if understand something wrongly. Some general comments,:
1. Generally it would be nice to know the ultimate goal of this work. As I understand this is just a step towards it, but I would like to know the whole scope.
2. It looks like you are not going to replace plan node tree with a new representation. Is that true? Or you just didn't yet implement that?
3. I have read somewhere that you are going to get rid of `Expression` from plan, am I right? I don't see it here, is this still on the road map? How are you going to achieve that.
4. Many vs single rule - it is not obvious to me that having many small rules instead of one big is beneficial. Take a look at `LimitPushDown` or `UnaliasSymbolReferences`. Legacy rules were huge, but now you will get plenty of simple small rules. In legacy all the contract between `visit*` methods were in the single file. Now it is distributed (all the push down limit rules push down limit node by one node). In case of adding new plan node, previously you had to implement a method, now you need to provide a class per each new plan node. Generally some rules are focused only on single node, but other would like to process whole tree. So I think it would be nice to have a choice. To be constructive, maybe `RuleGroup` would handle that.
```
class RuleGroup implements Rule {
  Optional<PlanNode> apply(...) {
     for (Rule rule : rules) {
        plan = rule.apply(...);
    }
  } 
}

class LimitPushDownRuleGroup {
  LimitPushDownRuleGroup() {
     rules = ImmutableList.of(
       new PushLimitOverSort(),
       ....);
  }
}
```

or to have giant IF rule:

```
class PushDownLimitRule implements Rule {
  Optional<PlanNode> apply(...) {
       if (node instanceof SortNode) {
         return rewriteSort(...);
      }
      ...
  }
} 
```
or some kind of visitor which understand the `lookup`.

5.  It is typical for rule that it is focused on single node, I think it would beneficial if `Rule` was registered for some node type in a kind of a hash map, so that way it would be only called for such nodes. No need to do `instanceof` in a `Rule::apply`.
6. Some rules can be applied only once (like mentioned limit push down), I wonder if it would be possible to somehow remove it from a list of rules so search space for the optimal plan could be limited. Generally I think we need a smarter way of applying rules as all to all may be to much resource consuming. What would you say for something like that: 
 - let garbage collect rules after each whole tree traversal. For example no apply node in a tree no need to run TransformApply* rules any more.
 - it would require to add an option to restore rules in case some other rule reintroduced apply node once again.
7. As I understand `GroupReference` is something like a pointer to a plan, so we could modify each plan node independently as long we don't mess with "pointers". Is it possible for plan node to be "pointed" (referenced) in multiple plan nodes? I think it would be nice to recognize common parts of plan, it would open a place for new optimizations. For example let say that `SELECT x,y,z FROM foo WHERE a AND b` appears several times in a plan (as a independent subqueries or side of joins) then we could leverage in memory connector to store a partial result there. etc.
8. It would be nice to automatically discover circuits between rules. Where rule A changes a plan that rule B can be applied, and rule B changes a plan that rule A can be applied.
9. As I understand `Memo` can grow without any limits, I think it should store only nodes which can be still accessed from the `rootGroup`. It requires some garbage collecting.
10. I think it would be nicie to leverage your code for cost based optimizer, so smarter rule applying and memo memory efficiency seem to be very important. Can write about how do you envision cost based optimizer with this?
11. This does not affect semantic analyzer nor regular planner (`QueryPlanner`, `RelationPlanner`). Is this intentional? Are you going to change planner? If not the https://github.com/prestodb/presto/issues/6625 seems to be still valid.

From my point of view question regarding scope and cost based optimizer are the most important. Even though you have implemented a lot, still is a lot to do. Would you accept any help? I think there are a lot Teradaters which would like to join this effort (including me of course).

kokosing commented:
12. The way you migrate optimizers blocks to have `IterativeOptimzer` be hidden behind feature toggle. As limit push down (or unalias) optimization is either implemented in legacy or new optimizer. In order to keep legacy optimizer for a while I think we need a period when rules are implemented in both optimizers. If you migrate a rule to only new optimizer then new optimizer is mandatory to be executed during query processing.

kokosing commented:
Is https://github.com/prestodb/presto/wiki/New-Optimizer up to date? Is this pull request part of it?

martint commented:
> Generally it would be nice to know the ultimate goal of this work. As I understand this is just a step towards it, but I would like to know the whole scope.

It's a first step towards revamping the optimizer and bring some sanity to how we're doing things. The biggest challenge is doing it incrementally.

> It looks like you are not going to replace plan node tree with a new representation. Is that true? Or you just didn't yet implement that?
I have read somewhere that you are going to get rid of Expression from plan, am I right? I don't see it here, is this still on the road map? How are you going to achieve that.

Not in this change. I'm trying to get some of the ideas I've been thinking about into 

> Many vs single rule - it is not obvious to me that having many small rules instead of one big is beneficial. Take a look at LimitPushDown or UnaliasSymbolReferences. Legacy rules were huge, but now you will get plenty of simple small rules. In legacy all the contract between visit* methods were in the single file. Now it is distributed (all the push down limit rules push down limit node by one node). In case of adding new plan node, previously you had to implement a method, now you need to provide a class per each new plan node. Generally some rules are focused only on single node, but other would like to process whole tree. So I think it would be nice to have a choice. To be constructive, maybe RuleGroup would handle that.

Possibly. I see this flexibility as a major benefit. The current plan node hierarchy is too rigid (in part, due to the limitations imposed by the visitor interface), and it couples the transformation rules with the exploration strategy. One of the benefits of the new approach is that it doesn't require every visitXXX method to be implemented for a rule to be useful. Today, if one is missing, the whole subtree of a node of that type doesn't get optimized. It also forces implementers to think about how a given transformation applies to every single node.

> It is typical for rule that it is focused on single node, I think it would beneficial if Rule was registered for some node type in a kind of a hash map, so that way it would be only called for such nodes. No need to do instanceof in a Rule::apply.

Yes, what I have is just the first step. My goal is to let rules specify a pattern the match on (e.g., like a regex on the tree structure, similar to what the plan assertion DSL can do) and take advantage of that information in the internal data structure the optimizer keeps.

> Some rules can be applied only once (like mentioned limit push down), I wonder if it would be possible to somehow remove it from a list of rules so search space for the optimal plan could be limited. 

Yeah, that's something I've also been considering. It will become more important once we switch to a comprehensive exploration optimizer. I'm not going to optimize for that case until we need it, though.

> As I understand GroupReference is something like a pointer to a plan, so we could modify each plan node independently as long we don't mess with "pointers". Is it possible for plan node to be "pointed" (referenced) in multiple plan nodes? I think it would be nice to recognize common parts of plan, it would open a place for new optimizations. For example let say that SELECT x,y,z FROM foo WHERE a AND b appears several times in a plan (as a independent subqueries or side of joins) then we could leverage in memory connector to store a partial result there. etc.

Indeed. The new plan node hierarchy will hopefully make that easier to do (some experiments in https://github.com/martint/presto/tree/optimizer). It requires stricter semantics for equality and hashcode and plan structure, so it's not easy to do with the current hierarchy.
 
> It would be nice to automatically discover circuits between rules. Where rule A changes a plan that rule B can be applied, and rule B changes a plan that rule A can be applied.

Yup, some of these behaviors will be easier to detect and handle in the future (but see the point above).

> As I understand Memo can grow without any limits, I think it should store only nodes which can be still accessed from the rootGroup. It requires some garbage collecting.

I have a TODO in IterativeOptimizer to do exactly that.

> I think it would be nicie to leverage your code for cost based optimizer, so smarter rule applying and memo memory efficiency seem to be very important. Can write about how do you envision cost based optimizer with this?

That's a longer term discussion. There are some ideas in the wiki page you linked to below. Matt, Akshat and I were talking about getting a bunch of people together to get a brain dump from me.

> This does not affect semantic analyzer nor regular planner (QueryPlanner, RelationPlanner). Is this intentional? Are you going to change planner? If not the #6625 seems to be still valid.

Not yet, but that's the eventual plan.

> The way you migrate optimizers blocks to have IterativeOptimzer be hidden behind feature toggle. As limit push down (or unalias) optimization is either implemented in legacy or new optimizer. In order to keep legacy optimizer for a while I think we need a period when rules are implemented in both optimizers. If you migrate a rule to only new optimizer then new optimizer is mandatory to be executed during query processing.

Yes, I'm currently working on that.

> Is https://github.com/prestodb/presto/wiki/New-Optimizer up to date? Is this pull request part of it?

I haven't updated it in a while. As I mentioned above, this is related, but it's not the implementation described in that wiki. It borrows some ideas (iterative optimizer and transformation rules), but it doens't, yet, replace the plan node hierarchy or implement memoization, cost, pattern matching, etc.

kokosing commented:
>> The way you migrate optimizers blocks to have IterativeOptimzer be hidden behind feature toggle. As limit push down (or unalias) optimization is either implemented in legacy or new optimizer. In order to keep legacy optimizer for a while I think we need a period when rules are implemented in both optimizers. If you migrate a rule to only new optimizer then new optimizer is mandatory to be executed during query processing.

> Yes, I'm currently working on that.

I would copy paste `PlanOptimizers` as `NewPlanOptimizers` and migrate only optimizers from the new one. Then I would add a feature toggle which would decide which optimizers should be used. That way you could merge your code sooner with a number of smaller PR (vs one huge like this one).  Legacy optimizer could removed once new ones get stable.

---
Generally it looks we all agree ;) It seems that optimizer is no longer a rocket science in database world, though it is not an easy part of it.

I think it would be really great to create aggregation github issue where you could put a new optimizer milestones (a roadmap) with coarse-grained points like: introduce iterative optimizer, rules migration, legacy optimizer removal. Each of them could have a list of smaller issues. The closes milestone could be detailed the farthest could be vague. Thanks to that we could work on independent features in parallel together. One could migrate rules, one could implement garbage collector in `Memo` etc. IMHO new optimizer is too much of work for a single person to make this happen in realistic term.

martint commented:
@dain, commits not marked "WIP" are ready for review. They are mostly independent from each other, and can be reviewed and merged in the order in which they show up.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+            for (Rule rule : rules) {
+                long start = System.nanoTime();
+                Optional<PlanNode> transformed = rule.apply(node, lookup, idAllocator, symbolAllocator);
+                stats.get(rule).timeNanos += System.nanoTime() - start;
```
I'm removing the stats code. I added it while I was working on the optimizer to see what the rules were doing. I'll re-introduce it later and expose the stats via jmx.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnaliasProject.java:
```diff
@@ -0,0 +1,99 @@
+import static com.facebook.presto.sql.planner.iterative.rule.Util.renameOutputs;
+
+/**
+ * Given:
```
Maybe we should add a sentence to each of these describing what and why we want the optimization. Something like:
```
Summary: Eliminate simply symbol renames in projections, which makes plan easier to read.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnaliasProject.java:
```diff
@@ -0,0 +1,99 @@
+
+        ProjectNode child = (ProjectNode) source;
+
+        Map<Symbol, Symbol> renames = Util.getRenames(child);
```
static import?

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnaliasProject.java:
```diff
@@ -0,0 +1,99 @@
+                                .entrySet().stream()
+                                .collect(Collectors.toMap(
+                                        Map.Entry::getKey,
+                                        e -> Util.rename(e.getValue(), renames)))));
```
static import

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java:
```diff
@@ -0,0 +1,76 @@
+            public Expression rewriteSymbolReference(SymbolReference reference, Object context, ExpressionTreeRewriter treeRewriter)
+            {
+                Symbol symbol = Symbol.from(reference);
+                return renames.getOrDefault(symbol, symbol).toSymbolReference();
```
you could call the `rename` function above

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnaliasFilter.java:
```diff
@@ -0,0 +1,107 @@
+
+        ProjectNode child = (ProjectNode) source;
+
+        Map<Symbol, Symbol> renames = Util.getRenames(child);
```
same thing about static imports

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java:
```diff
@@ -0,0 +1,169 @@
+
+    private int nextGroupId;
+
+    private final Map<Integer, PlanNode> membership = new HashMap<>();
```
Is this going to evolve that eventually a _set_ of semantically equivalent PlanNodes will be assigned to groupId?

martint replied:
It will probably be a completely different implementation of Memo. There are some challenging edge case due to cycles, etc

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+                if (transformed.isPresent()) {
+                    stats.get(rule).hits++;
+
+                    memo.transform(group, transformed.get());
```
IMO `replace`/`replaceGroup` or `substitute`/`substituteGroup` would be a better method name than `transform`.

martint replied:
Technically, "addEquivalence" would be more accurate. It's very likely that the Memo implementation will need to be designed to work hand-in-hand with the exploration algorithm, so that would be premature generalization.

`replace` seems reasonable. I'll change it to that.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+        boolean done = false;
+        boolean progress = false;
+
+        while (!done) {
```
some anti-looping mechanism is needed eventually.
E.g fail if exploring takes more than N seconds.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+        return progress;
+    }
+
+    private boolean exploreNode(
```
It seems like `node` arg can be dropped. The method is always called with `memo.getNode(group)`, and it can be done internally.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+                    memo.transform(group, transformed.get());
+
+                    // output node is special. TODO: we need to get rid of OutputNode. It doesn't add any value
+                    checkState(node instanceof OutputNode || new HashSet<>(node.getOutputSymbols()).equals(new HashSet<>(transformed.get().getOutputSymbols())),
```
This `checkState` should be part of `Memo.transform` method.

martint replied:
Not really. Memo.transform's job is just to record the transformation. This code is responsible for the application of rules. This check is just enforcing an invariant that rules need to satisfy

losipiuk replied:
But should Memo allow such transformation? 
You may treat it as a totally dumb structure which does not understand semantics of nodes at all. But I am not sure if that is truly beneficial. 

We are gaining some flexibility. But will we ever use that? 
On the other hand you you are making Memo api more prone to misusage. As every caller of Memo.transform have to take care of constraints with code similar to commented `checkState`.

martint replied:
You have a good point. It's a good argument that Memo should not allow replacing an expression with one that doesn't produce the same outputs (an fail with a IllegalArgumentException or something). I'll update it.

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+
+        while (!done) {
+            done = true;
+            for (Rule rule : rules) {
```
Note: We probably want a less trivial queue here, so we prioritise peeking rules which are expected to actually make some changes, to avoid unproductive iterations.

losipiuk commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java:
```diff
@@ -0,0 +1,256 @@
+
+import static org.testng.Assert.assertEquals;
+
+public class TestMemo
```
Too much vertical whitespace

martint replied:
Can you elaborate?

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSampleAsFilter.java:
```diff
@@ -0,0 +1,72 @@
+ *     - X
+ * </pre>
+ */
+public class ImplementSampleAsFilter
```
This should probably be call ImplementBernoulliSampleAsFilter

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnaliasProject.java:
```diff
@@ -0,0 +1,99 @@
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
```
All the `Unalias*` rules have very similar structure.
Maybe it would be better to actually have single rule handling all the nodes, or common supertype for all those rules?
Then you would be able to get rid of ugly `Util` helper class.

martint replied:
I've gone back and forth on this. Having a single rule to handle all of them makes it harder to read, and reason about. Another thing is that, eventually, I want connectors to be able to participate in the optimization process by providing rules which might return connector-specific node types. They will also need to provide rules to do something similar to this, which brings us back to having multiple rules with similar structure or that need to reuse common "utility" functionality.

Some of the common structure (and this applies to pretty much every rule) has to do with checking whether the rule should fire. Today, this has to be done manually. In the future, it will be possible for rules to describe a pattern that the tree needs to match before the rule is called.

losipiuk replied:
Sure. Let's leave it. And see what is needed eventually.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,178 @@
+
+        exploreGroup(memo, lookup, idAllocator, symbolAllocator, memo.getRootGroup());
+
+        if (debug) {
```
I'm removing this for now

martint commented:
I merged the first commit ("move canonicalizer..."), addressed feedback and removed all the WIP commits so that we can get a proper run in travis.

BTW, I got rid of the Util class and introduce a Renamer, instead, since it was mostly a bunch of utility methods for that.

kokosing commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -271,6 +272,11 @@ public SystemSessionProperties(
+                        false),
+                booleanSessionProperty(
+                        NEW_OPTIMIZER,
+                        "Experimental: enable new optimizer",
```
`new` does not tell much. Maybe use `iterative`?

kokosing replied:
also you can extract this feature toggle as separate commit.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,129 @@
+        return memo.extract();
+    }
+
+    private boolean exploreGroup(Memo memo, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, int group)
```
Instead of passing session objects (memo, lookup, idAllocator, symbolAllocator) in all the methods I would extract separate class which could have these objects as fields

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java:
```diff
@@ -0,0 +1,129 @@
+
+        Memo memo = new Memo(idAllocator, plan);
+
+        Lookup lookup = node -> {
```
As I understand the aim of this `Lookup` is hide `memo` from rules. If it is true then maybe `memo` itself could provide such function object `memo.getLookup()` ?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java:
```diff
@@ -0,0 +1,176 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+public class Memo
```
I would extract plan extraction from this class to a separate class which could leverage the `Lookup` function. I think the same could apple for building. This class is going to be much more complicated in the future, so I would prefer to keep separate concepts in separate classes. Up to you.

martint replied:
will do

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java:
```diff
@@ -0,0 +1,176 @@
+    private int nextGroupId;
+
+    private final Map<Integer, PlanNode> membership = new HashMap<>();
+    private final Map<Integer, Integer> referenceCounts = new HashMap<>();
```
what is the purpose of this map? Is it used for garbage collection? Then is not mark and sweep a better approach for collection?

kokosing replied:
Since it is not yet used (other updating the state) I would remove it for now.

martint replied:
> what is the purpose of this map? Is it used for garbage collection? Then is not mark and sweep a better approach for collection?

Yes, it's for GC. It's used below (see incrementReferenceCounts and decrementReferenceCounts).

Mark & sweep is another alternative (and I considered it), but it raises questions about how often to do it, how big to let the memo grow, etc. Since there can be no cycles in the memo structure, it's also overkill.

> Couldn't you use java weak references to just remove groups for which there is no GroupReference? You would need some sort of GroupId, but that should be fine.

I considered this, too, but I don't think it's that much easier. It poses other problems: GroupId instance identity would matter, so we'd need to make sure they are never created outside of some very controlled scope. More importantly, if someone outside of the Memo held on to one of them, it would cause a memory leak.

> In what case it is good to remove group? Won't this greatly limit your plan search space? E.g: it might be useful to leave group so that some other rules produce even better plan starting from to be-deleted group.

This memo implementation only keeps around one expression per group. So, effectively, it holds on to just one plan. Keeping track of multiple plans will require a different implementation (I've prototyped some of this in my other branch, mentioned in https://github.com/prestodb/presto/wiki/New-Optimizer).

kokosing replied:
> Mark & sweep is another alternative (and I considered it), but it raises questions about how often to do it, how big to let the memo grow, etc. Since there can be no cycles in the memo structure, it's also overkill.

Mark & sweep has one advantage I like over reference counting. It can be extracted to separate class (if memo expose API that can remove group) or method. Thanks to that:
 * you don't need to care about updating reference counts
 * memo itself is simpler
 * in case of different implementation of memo you can simply reuse it (vs extending some abstract class etc or reimplementing garbage collection)

I think we could start running mark and sweep once per application of all the rules for the root node.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java:
```diff
@@ -0,0 +1,176 @@
+    {
+        PlanNode old = membership.get(group);
+
+        // output node is special. TODO: we need to get rid of OutputNode. It doesn't add any value
```
Sometimes it may a value to keep such nodes as `OutputNode`. For example in case when optimizer optimizes a batch of queries (not just a single query) then it might use such dummy node to group queries together.

Just a note. 

Anyway I would try to handle `OutputNode` as any other node, it is just a special `ProjectNode`.

kokosing replied:
Given the explanation below (the one for checkArgument) I do not understand why `OutputNode` is so special. It is a root node and it output symbols have to not to change. Maybe `equals` usage is to strict? What about checking only types and count of output symbols (but not names)? Would it help to remove this exception case for `OutputNode`?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java:
```diff
@@ -0,0 +1,176 @@
+        PlanNode old = membership.get(group);
+
+        // output node is special. TODO: we need to get rid of OutputNode. It doesn't add any value
+        checkArgument(node instanceof OutputNode || new HashSet<>(old.getOutputSymbols()).equals(new HashSet<>(node.getOutputSymbols())),
```
what about `PruneUnreferencedOutputs`? This rule will break this check. Are you going to port this rule?

martint replied:
Actually, no :)

The way to model PruneXXX rules is to look for, say, a subplan with the following shape:

```
- X
   - Y => [f1, ..., fn]
      - Z
```

Then, determine, whether the columns referenced by expressions in X are a subset of the outputs of Y. If so (say, in this example, only columns f1, f2, f3 are used), rewrite the plan subplan as:

```
- X
   - Y' => [f1, f2, f3]
      - Z
```

With any expression that produces f4, ..., fn in Y removed.

So, X would continue to produce the same outputs. This is a key aspect of the contract for rules and groups in the memo. Rules transform subplans into equivalent subplans, where "equivalent" means, rougly, "they produce the same (multi)set of tuples".

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java:
```diff
@@ -0,0 +1,176 @@
+        return ChildReplacer.replaceChildren(node, children);
+    }
+
+    public void replace(int group, PlanNode node, String rule)
```
s/node/newNode ?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java:
```diff
@@ -0,0 +1,25 @@
+
+public interface Rule
+{
+    Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator);
```
I wonder if there is a case when `Rule` might need access to `Memo`. 

For example a rule which tries to identify common subplans and replace each of their occurrence with the same `GroupNode`, then such rule needs an access to the whole plan. Do you think that such rule should recursively call `lookup`  in order to traverse the plan in the current design?

sopel39 replied:
Right, it seem that `Rule` might often need more than just `Lookup` translator. E.g: calcite provides rule application context: https://calcite.apache.org/apidocs/org/apache/calcite/plan/RelOptRuleCall.html, which has many utility methods used by rules: `getParents()`, `transformTo`, etc. This interface be refactored though in the future.

martint replied:
> I wonder if there is a case when Rule might need access to Memo.

Not directly. I want to avoid a dependency between Rule <-> Memo for a couple of reasons:
- Eventually, connectors will be able to provide rules in addition to functions, types, tables, etc.
- The cost-based optimizer will need a different implementation of the Memo with slightly different semantics.

martint replied:
> Right, it seem that Rule might often need more than just Lookup translator

Indeed. One thing we'll probably need is for rules to be able to get certain "traits" out of plan expressions (e.g., effective predicate, sortedness, functional dependencies, etc). Most of these will probably need to be computed on-demand and cached for each node.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java:
```diff
@@ -0,0 +1,250 @@
+        return new ValuesNode(id, ImmutableList.of(), ImmutableList.of());
+    }
+
+    private ValuesNode values()
```
what about putting this methods like `values` or `filter` to separate class like `PlanTestDsl`?

martint replied:
It's not a general purpose dsl. It's just for convenience to avoid sprinkling a bunch of `new ProjectNode(...)`, etc. in every test in this class. Also, I don't really care about it being filters or projections or values, since the purpose is just to test whether the Memo is breaking up and reassembly the plans correctly. If the plan node hierarchy weren't so rigid, I would've created specialized node types. Hmm.. now that I think about it, I may be able to do just that anyway, since the code doesn't use visitors.

martint replied:
Spoke too soon. It does need visitors for ChildReplacer.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java:
```diff
@@ -0,0 +1,250 @@
+                        project(y2.getId(), values(z.getId()))));
+    }
+
+    private static void assertMatchesStructure(PlanNode actual, PlanNode expected)
```
I think it would be better to use `PlanMatchPattern` here. I don't think it is a good idea to create two frameworks for the same.

martint replied:
It's not that simple, and that's probably overkill. For this, I only care about matching just the overall structure and I'm not concerned about the internals of each node. One thing it needs to verify, in particular is that we get the node ids we expect.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -79,7 +79,9 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
         builder.add(
```
can you add an instance of `AbstactTestQueries` which is using new optimizer to unit tests?

kokosing replied:
maybe as separate commit between this one and the one before.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementFilteredAggregations.java:
```diff
@@ -0,0 +1,122 @@
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
```
I think there will be a long time before legacy optimizers got deleted. The problem I see is that once a rule is migrated then only legacy optimizers will get updated (bugfixes, new features). 

Generally code here is copy pasted from the other rule. I wonder if it possible to somehow share the code for time being so new rules got updates also. This would probably require change the legacy rule to somehow match new optimizer interface:
```
@Override
        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Optional<Symbol>> context)
        {
            Optional<PlanNode> transformed = ...;
            if (! transformed.isPresent()) {
                return context.defaultRewrite(node, Optional.empty());
```

so the code under `...` could be somehow used here.

What do you think?

martint replied:
It's hard to do in general. Most of the current optimizers have the "precondition", traversal and core logic all entangled. The precondition needs to be fairly cheap, as it will be evaluated multiple times during the optimization process.

dain commented:
@martint if this needs another review let me know

findepi commented:
@martint can we close this?

kokosing commented:
@martint Can we close this?

martint commented:
Yeah, let's close this now. Although some of the rules still need to be migrated, most of the changes are already in.
